### PR TITLE
Paper

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

JOSS paper draft

### CODE REVIEW

File: paper.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+As part of the Qiskit Ecosystem (https://qiskit.org/ecosystem), the package interfaces with other parts of Qiskit `@Qiskit`. Most notably, Qiskit Dynamics provides tools for simulating control sequences specified by Qiskit Pulse `@alexander_qiskit_2020`, which is used to specify hardware-level control of quantum computers. Higher level interfaces allow users to build and interact with simulation-based objects mimicking real IBM Quantum computers. (This last sentence is weird, having trouble summarizing the idea of ``DynamicsBackend`` without defining "backend".)
```
Comment (ihincks) [inline paper.md:60]:
```suggestion
As part of the Qiskit Ecosystem (https://qiskit.org/ecosystem), the package interfaces with other parts of Qiskit `@Qiskit`. Most notably, Qiskit Dynamics provides tools for simulating control sequences specified by Qiskit Pulse `@alexander_qiskit_2020`, which is used to specify hardware-level control of quantum computers. Higher level interfaces allow users to build and interact with simulation-based objects that target the same constraints (coupling map, timing, etc.) as a specified IBM Quantum computer. (This last sentence is weird, having trouble summarizing the idea of ``DynamicsBackend`` without defining "backend".)
```

File: paper.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+   index: 2
+ - name: IBM Quantum, IBM Research Tokyo, Tokyo, Japan
+   index: 3
+date: 31 July 2023
+bibliography: paper.bib
+---
+
+# Summary
+
+Qiskit Dynamics is an open source Python library for numerically simulating the time dynamics of finite-dimensional quantum systems. The goal of the package is to provide flexible configuration of the numerical methods used for simulation: choice of array representations (dense v.s. sparse, and different array libraries), access to different types of underlying solvers (standard ODE v.s. geometric solvers), as well as general tools for transforming models of quantum systems for more efficient simulation (rotating frames and the rotating wave approximation). The package also contains advanced functionality for computing time-dependent perturbation theory expressions used in robust quantum control optimization `@perturb1`, `@perturb2`.
```
Comment (ihincks) [inline paper.md:58]: Feel free to ignore, but I'd put the (imo) "most interesting" feature first.
```suggestion
Qiskit Dynamics is an open source Python library for numerically simulating the time dynamics of finite-dimensional quantum systems. The goal of the package is to provide flexible configuration of the numerical methods used for simulation: general tools for transforming models of quantum systems for more efficient simulation (rotating frames and the rotating wave approximation), choice of array representations (dense v.s. sparse, and different array libraries), and access to different types of underlying solvers (standard ODE v.s. geometric solvers). The package also contains advanced functionality for computing time-dependent perturbation theory expressions used in robust quantum control optimization `@perturb1`, `@perturb2`. 
```

**Review by mtreinish — APPROVED** [review]
LGTM

File: paper.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+   index: 9
+
+
+date: 31 July 2023
+bibliography: paper.bib
+---
+
+# Summary
+
+Qiskit Dynamics is an open source Python library for numerically simulating the time dynamics of finite-dimensional quantum systems. The goal of the package is to provide flexible configuration of the numerical methods used for simulation: general tools for transforming models of quantum systems for more efficient simulation (rotating frames and the rotating wave approximation), choice of array representations (dense v.s. sparse, and different array libraries), and access to different types of underlying solvers (standard ODE v.s. geometric solvers). The package also contains advanced functionality for computing time-dependent perturbation theory expressions used in robust quantum control optimization `[@perturb1; @perturb2]`.
```
Comment (eggerdj) [inline paper.md:60]: I wonder if this summary is a too heavy on the jargon? E.g. rotating frame
```suggestion
Qiskit Dynamics is an open source Python library to numerically simulate the time dynamics of finite-dimensional quantum systems. The package provides a flexible configuration of the numerical methods used for the simulation. Users can chose how to transform the models of the quantum systems to enable efficient simulations, how to represent the underlying arrays (e.g., dense v.s. sparse with different array libraries), and with which solvers to time-evolve the models (e.g., standard ODE v.s. geometric solvers). Qiskit Dynamics also contains advanced functionality to compute time-dependent perturbation theory expressions used in robust quantum control optimization `[@perturb1; @perturb2]`. 
```

File: paper.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+As part of the Qiskit Ecosystem (https://qiskit.org/ecosystem), the package interfaces with other parts of Qiskit `[@Qiskit]`. Most notably, Qiskit Dynamics provides tools for simulating control sequences specified by Qiskit Pulse `[@alexander_qiskit_2020]`, which is used to specify hardware-level control of quantum computers. Higher level interfaces allow users to build and interact with simulation-based objects that target the same constraints (coupling map, timing, etc.) as a specified IBM Quantum computer.
```
Comment (eggerdj) [inline paper.md:62]:
```suggestion
As part of the Qiskit Ecosystem (https://qiskit.org/ecosystem), the package interfaces with other parts of Qiskit `[@Qiskit]`. Most notably, Qiskit Dynamics provides tools for simulating control sequences specified by Qiskit Pulse `[@alexander_qiskit_2020]` which gives a pulse representation of quantum circuit instructions. Higher level interfaces allow users to build and interact with simulation-based objects that target the same constraints (coupling map, timing, etc.) as a specified IBM Quantum computer.
```
I find the last sentence a bit hard to follow but don't have a concrete suggestion. What is meant here: `Higher level interfaces allow users to build and interact with simulation-based objects that target the same constraints (coupling map, timing, etc.) as a specified IBM Quantum computer.`

File: paper.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+Lastly, to facilitate high-perfomance applications, Qiskit Dynamics is compatible with the JAX array library `[@jax2018github]`. As such, all core computations are just-in-time compilable, automatically differentiable, and executable on GPU.
```
Comment (eggerdj) [inline paper.md:64]: Maybe state that this provides a big boost to speed?

File: paper.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+# Statement of need
```
Comment (eggerdj) [inline paper.md:66]: If I were a ref I would ask: Don't we already have all this in Qutip? Why do we need Qiskit Dyamics?

I think you could make a case around the Qiskit Pulse compatibility and saying that Qiskit Dynamics also naturally integrates in the Qiskit ecosystem. Or sth along those lines.

File: paper.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+Furthermore, having a simulation-based drop-in replacement for real quantum computing systems is useful for developers building software tools for low-level control of experiments, such as Qiskit Pulse `[@alexander_qiskit_2020]` and Qiskit Experiments `[@kanazawa_qiskit_2023]`.
+
+# Related open source packages
+
+Due to its importance, many open source packages contain time-dependent quantum system simulation tools. In Python, these include QuTiP `[@qutip]`, TorchQuantum `[@torchquantum]`, and C3 `[@C3]`. C++ packages (also with Python interfaces) include lindbladmpo `[@lindbladmpo]` and Quandary `[@quandary]`. Packages also exist in other languages, such as the Hamiltonian open quantum system toolkit (HOQST) `[@hoqst]` and a Framework for Quantum Optimal Control `[@julia_qc]` in Julia, and Spinach `[@spinach]` in MATLAB.
```
Comment (eggerdj) [inline paper.md:74]: A few words on how we differentiate?

File: paper.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+# Documentation and community
```
Comment (eggerdj) [inline paper.md:76]: Do you have a cool eye-catching figure to add? Could be nice.

File: paper.md (hunk 4 above)
Comment (DanPuzzuoli) [reply paper.md:62]: Basically we're trying to describe the `DynamicsBackend` without having to explain what a `Backend` is.

Comment (DanPuzzuoli) [reply paper.md]: Maybe something like:

```
By specifying a physical model for a quantum computer, users can construct an object imitating a ``backend'' object in Qiskit, which is the interface used to submit jobs to real IBM quantum computers. Using this object, users can submit jobs (in the form of quantum circuits or Qiskit Pulse schedules) and receive simulation-based results in the same format as they would with a real IBM quantum computer.
```

Comment (eggerdj) [reply paper.md]: Ok I see. Here is a suggestion. I think it can be improved, perhaps it gives some inspiration?
```
This enables one to build a simulator of a quantum computer that time-evolves user provided pulses, encapsulated in quantum circuit instructions, according to a Hamiltonian describing the computer. This representation is closer to the physical dynamics of the quantum computer than a simulator based on gate-level unitary matrices.
```

File: paper.md (hunk 6 above)
Comment (DanPuzzuoli) [reply paper.md:66]: After reading the journal requirements and examples, I don't actually think there is any need to justify why a new package was written v.s. using something pre-existing. The guidelines for authors, as far as I can tell, don't say anything about explaining your package within the context of other work (aside from citing it). I don't mean this as a denial that this is a generally important point; just explaining why I wrote this section in this way.

So, with this in mind, I wrote the "statement of need" section as a simple description of "why do we need these kinds of simulations/interfaces", rather than a "why do we need these kinds of simulations/interfaces AND why did we write another package". It may be high level, but everything I say in that paragraph does implicitly explain why Dynamics was written v.s. just using qutip. I do also describe qiskit integration in the second paragraph of this section. 

What are your thoughts on this? We could rearrange it to highlight the qiskit integration first, as it is the most obviously unique feature, then describe the other technical stuff after.

File: paper.md (hunk 7 above)
Comment (DanPuzzuoli) [reply paper.md:74]: I assume here you mean "how do we compare"? My response here is roughly the same as in the previous point: is there any need to make comparisons?

File: paper.md (hunk 8 above)
Comment (DanPuzzuoli) [reply paper.md:76]: I don't :).

File: paper.md (hunk 3 above)
Comment (DanPuzzuoli) [reply paper.md:60]: Yeah I was on the fence about including "rotating frame" in the text. As a compromise I ended up putting it in parentheses, as I'd still like to be specific for readers that know what these things are.

File: paper.md (hunk 5 above)
Comment (DanPuzzuoli) [reply paper.md:64]: I was trying to use "high-perfomance" to capture this :). I can rearrange this though a bit to be more explicit about speed improvements.

File: paper.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+As part of the Qiskit Ecosystem (https://qiskit.org/ecosystem), the package interfaces with other parts of Qiskit `[@Qiskit]`. Most notably, Qiskit Dynamics provides tools for simulating control sequences specified by Qiskit Pulse `[@alexander_qiskit_2020]` which gives a pulse representation of quantum circuit instructions. Higher level interfaces allow users to build and interact with simulation-based objects that target the same constraints (coupling map, timing, etc.) as a specified IBM Quantum computer.
```
Comment (to24toro) [inline paper.md:42]: What does `timing` indicate?
I find that the name `coupling map` may not be familiar to general users. How about deleting it or adding an explanation?

Comment (DanPuzzuoli) [reply paper.md:42]: Timing indicates the `dt` for the backend. I think I'd still like to keep this to elaborate on the what "constraints" are, but maybe some other more colloquial words would be better. E.g. "qubit layout" instead of "coupling map", and "sampling rate" instead of "timing".

Comment (DanPuzzuoli) [reply paper.md]: @to24toro I've just edited this, let me know what you think.

Comment (to24toro) [reply paper.md]: Thanks, @DanPuzzuoli .

LGTM!

**Review by to24toro — APPROVED** [review]
